blockbuilder/blockbuilder.go

```diff
@@ -0,0 +1,249 @@
+	txns []*types.Transaction
+
+	// block is a reference to the already built block
+	block *types.Block
+
+	state *state.Transition
+}
+
+func (b *BlockBuilder) Reset() error {
+	// Generate the base block
+	header := &types.Header{
+		ParentHash: b.params.Parent.Hash,
+		Number:     b.params.Parent.Number + 1,
+		Miner:      types.ZeroAddress.Bytes(),
+		Nonce:      types.Nonce{},
+		MixHash:    signer.IstanbulDigest,
+		Difficulty: b.params.Parent.Number + 1,
+		StateRoot:  types.EmptyRootHash,
+		Sha3Uncles: types.EmptyUncleHash,
+		GasLimit:   b.params.Parent.GasLimit,
+	}
+
+	b.block = nil
+	b.txns = []*types.Transaction{}
+
+	// calculate gas limit based on parent header
+	header.GasLimit = calculateGasLimit(b.params.Parent.GasLimit, b.params.BlockGasTarget)
+
+	transition, err := b.params.Executor.BeginTxn(b.params.Parent.StateRoot, b.header, b.params.Coinbase)
+	if err != nil {
+		return err
+	}
+
+	b.state = transition
+
+	return nil
+}
+
+// Fill fills the block with transactions from the txpool
+func (b *BlockBuilder) Fill() error {
+	b.params.TxPool.Prepare()
+
+write:
+	for {
+		tx := b.params.TxPool.Peek()
+
+		// execute transactions one by one
+		finished, err := b.writeTxPoolTransaction(tx)
+		if err != nil {
+			b.params.Logger.Debug("Fill transaction error", "hash", tx.Hash, "err", err)
+		}
+
+		if finished {
+			break write
+		}
+	}
+
+	return nil
+}
+
+func (b *BlockBuilder) writeTx(tx *types.Transaction) error {
+	if err := b.state.Write(tx); err != nil {
+		return err
+	}
+
+	b.txns = append(b.txns, tx)
+
+	return nil
+}
+
+func (b *BlockBuilder) writeTxPoolTransaction(tx *types.Transaction) (bool, error) {
+	if tx == nil {
+		return true, nil
+	}
+
+	if err := b.writeTx(tx); err != nil {
+		if _, ok := err.(*state.GasLimitReachedTransitionApplicationError); ok { //nolint:errorlint
+			// stop processing
+			return true, err
+		} else if appErr, ok := err.(*state.TransitionApplicationError); ok && appErr.IsRecoverable { //nolint:errorlint
+			b.params.TxPool.Demote(tx)
+
+			return false, err
+		} else {
+			b.params.TxPool.Drop(tx)
+
+			return false, err
+		}
+	}
+
+	// remove tx from the pool and add it to the list of all block transactions
+	b.params.TxPool.Pop(tx)
+
+	return false, nil
+}
+
+func (b *BlockBuilder) Build(handler func(*types.Header)) *StateBlock {
+	// Commit the changes
+	_, root := b.state.Commit()
+
+	// Update the header
+	b.header.StateRoot = root
+	b.header.GasUsed = b.state.TotalGas()
+
+	if handler != nil {
+		handler(b.header)
+	}
+
+	// Build the actual block
+	// The header hash is computed inside buildBlock
+	b.block = consensus.BuildBlock(consensus.BuildBlockParams{
+		Header: b.header,
+		Txns:   b.txns,
+		// TODO: Capture the receipts from `state.Write`
+		Receipts: b.state.Receipts(),
+	})
+
+	b.block.Header.ComputeHash()
+
+	return &StateBlock{
+		Block:    b.block,
+		Receipts: b.state.Receipts(),
+	}
+}
+
+// StateBlock is a block with the full state it modifies
```

Comment:
I think we can remove `StateBlock`, receipts can be retrieved from `types.Block`